core/src/main/java/org/infinispan/persistence/async/AsyncNonBlockingStore.java

```diff
@@ -0,0 +1,758 @@
+               completeExceptionally(throwable);
+            } else {
+               complete(null);
+            }
+         });
+      }
+   }
+
+   private CompletionStage<Void> replicateModifications(Map<Object, Modification> modifications) {
+      // Use a connected flowable, so we don't have to iterate over the modifications twice
+      ConnectableFlowable<Modification> connectableModifications = Flowable.fromIterable(modifications.values())
+            .publish();
+
+      // The below two methods may lazily subscribe to the Flowable, thus we must auto connect after both are
+      // subscribed to (e.g. NonBlockingStoreAdapter subscribes on a blocking thread)
+      Flowable<Modification> modificationFlowable = connectableModifications.autoConnect(2);
+
+      CompletionStage<Void> putStage = actual.bulkWrite(segmentCount, modificationFlowable.ofType(PutModification.class)
+            .groupBy(Modification::getSegment, PutModification::<K, V>getEntry)
+            .map(SegmentPublisherWrapper::new));
+      CompletionStage<Void> removeStage = actual.bulkDelete(segmentCount, modificationFlowable.ofType(RemoveModification.class)
+            .groupBy(Modification::getSegment, RemoveModification::getKey)
+            .map(SegmentPublisherWrapper::new));
+
+      return CompletionStages.allOf(removeStage, putStage);
+   }
+
+   private CompletionStage<Void> getAvailabilityDelayStage() {
+      if (asyncConfiguration.failSilently()) {
+         return CompletableFutures.completedNull();
+      }
+      CompletableFuture<Void> availabilityFuture;
+      synchronized (this) {
+         availabilityFuture = delegateAvailableFuture;
+      }
+      return availabilityFuture == null ? CompletableFutures.completedNull() : availabilityFuture;
+   }
+
+   @Override
+   public Publisher<MarshallableEntry<K, V>> publishEntries(IntSet segments, Predicate<? super K> filter, boolean includeValues) {
+      return Flowable.defer(() -> {
+         assertNotStopped();
+         if (trace) {
+            log.tracef("Publisher subscribed to retrieve entries for segments %s", segments);
+         }
+         return abstractPublish(segments, filter, PutModification::getEntry, MarshallableEntry::getKey,
+               (innerSegments, predicate) -> actual.publishEntries(innerSegments, predicate, includeValues));
+      });
+   }
+
+   @Override
+   public Publisher<K> publishKeys(IntSet segments, Predicate<? super K> filter) {
+      return Flowable.defer(() -> {
+         assertNotStopped();
+         if (trace) {
+            log.tracef("Publisher subscribed to retrieve keys for segments %s", segments);
+         }
+         return abstractPublish(segments, filter, putModification -> putModification.<K, Object>getEntry().getKey(),
+               RxJavaInterop.identityFunction(), actual::publishKeys);
+      });
+   }
+
+   private <E> Publisher<E> abstractPublish(IntSet segments, Predicate<? super K> filter, Function<PutModification, E> putFunction,
```

Comment:
I'm having some trouble following this method logic... but I was expecting to check RemoveModification to prevents those keys to be published.

Reply:
It should be. The `entryModifications` map will contain the RemoveModification, which isn't published in the `modPublisher` but is filtered via `combinedPredicate `